lib/Conversion/FIRRTLToHW/LowerToHW.cpp

```diff
@@ -2759,58 +2791,102 @@ LogicalResult FIRRTLLowering::visitStmt(StopOp op) {
+                                                     StringRef annoClass) {
+  Value clock, enable, predicate;
+  StringAttr label;
+  StringRef fileName, dir;
+  BOpTy svOp;
+  if (!lowerVerifOps<AOpTy>(op, clock, enable, predicate, label, fileName, dir,
+                            annoClass))
+    return failure();
+  predicate = builder.createOrFold<comb::AndOp>(enable, predicate);
+  sv::EventControl event;
+  if (op.eventControl() == "AtPosEdge")
+    event = circt::sv::EventControl::AtPosEdge;
+  else if (op.eventControl() == "AtEdge")
+    event = circt::sv::EventControl::AtEdge;
+  else if (op.eventControl() == "AtNegEdge")
+    event = circt::sv::EventControl::AtNegEdge;
+  else {
+    op.emitError(
+        "Event Control can only be one of :'AtPosEdge/AtEdge/AtNegEdge'");
+    return failure();
+  }
+  svOp = builder.create<BOpTy>(
+      circt::sv::EventControlAttr::get(builder.getContext(), event), clock,
+      predicate, label);
+
+  if (!fileName.empty() || !dir.empty())
+    svOp->setAttr("output_file",
+                  hw::OutputFileAttr::get(builder.getStringAttr(dir),
+                                          builder.getStringAttr(fileName),
+                                          builder.getBoolAttr(true),
+                                          builder.getBoolAttr(true),
+                                          svOp.getContext()));
   return success();
 }
 
 // Lower an assert to SystemVerilog.
 LogicalResult FIRRTLLowering::visitStmt(AssertOp op) {
-  return lowerVerificationStatement<AssertOp, sv::AssertOp>(op,
-                                                            assertAnnoClass);
+  return op.isConcurrent()
+             ? lowerVerificationConcurrentStatement<AssertOp,
+                                                    sv::AssertConcurrentOp>(
+                   op, assertAnnoClass)
+             : lowerVerificationStatement<AssertOp, sv::AssertOp>(
+                   op, assertAnnoClass);
 }
 
 // Lower an assume to SystemVerilog.
 LogicalResult FIRRTLLowering::visitStmt(AssumeOp op) {
-  return lowerVerificationStatement<AssumeOp, sv::AssumeOp>(op,
-                                                            assumeAnnoClass);
+  return op.isConcurrent()
+             ? lowerVerificationConcurrentStatement<AssumeOp,
+                                                    sv::AssumeConcurrentOp>(
+                   op, assumeAnnoClass)
+             : lowerVerificationStatement<AssumeOp, sv::AssumeOp>(
+                   op, assumeAnnoClass);
 }
 
 // Lower a cover to SystemVerilog.
 LogicalResult FIRRTLLowering::visitStmt(CoverOp op) {
-  return lowerVerificationStatement<CoverOp, sv::CoverOp>(op, coverAnnoClass);
+  return op.isConcurrent()
+             ? lowerVerificationConcurrentStatement<CoverOp,
+                                                    sv::CoverConcurrentOp>(
+                   op, coverAnnoClass)
+             : lowerVerificationStatement<CoverOp, sv::CoverOp>(op,
```

Comment:
Please merge lowerVerificationConcurrentStatement and lowerVerificationStatement to reduce the redundancy in each of the visit operations.

Reply:
Done, merged the functions by adding a third template argument.